echopype/echodata/utils_platform.py

```diff
@@ -0,0 +1,118 @@
+import numpy as np
+
+
+def _extvar_properties(ds, name):
+    """Test the external variable for presence and all-nan values,
+    and extract its time dimension name.
+    Returns <presence>, <valid values>, <time dim name>
+    """
+    if name in ds:
+        # Assumes the only dimension in the variable is a time dimension
+        time_dim_name = ds[name].dims[0] if len(ds[name].dims) > 0 else "scalar"
+        if not ds[name].isnull().all():
+            return True, True, time_dim_name
+        else:
+            return True, False, time_dim_name
+    else:
+        return False, False, None
+
+
+# clip incoming time to 1 less than min of EchoData["Sonar/Beam_group1"]["ping_time"] and
+#   1 greater than max of EchoData["Sonar/Beam_group1"]["ping_time"]
+# account for unsorted external time by checking whether each time value is between
+#   min and max ping_time instead of finding the 2 external times corresponding to the
+#   min and max ping_time and taking all the times between those indices
+def _clip_by_time_dim(external_ds, ext_time_dim_name, ping_time):
+    # external_ds is extra_platform_data[vars-list-on-time_dim_name]
+    sorted_external_time = external_ds[ext_time_dim_name].data
+    sorted_external_time.sort()
+    # fmt: off
+    min_index = max(
+        np.searchsorted(
+            sorted_external_time,
+            ping_time.min(),
+            side="left"
+        ) - 1,
+        0,
+    )
+    # fmt: on
+    max_index = min(
+        np.searchsorted(
+            sorted_external_time,
+            ping_time.max(),
+            side="right",
+        ),
+        len(sorted_external_time) - 1,
+    )
+    # TODO: this element-wise comparison is expensive and seems an ad-hoc patch
+    # to deal with potentially reversed timestamps in the external dataset.
+    # Review at workflow stage to see if to clean up timestamp reversals
+    # and just find start/end timestamp for slicing.
+    return external_ds.sel(
+        {
+            ext_time_dim_name: np.logical_and(
+                sorted_external_time[min_index] <= external_ds[ext_time_dim_name],
+                external_ds[ext_time_dim_name] <= sorted_external_time[max_index],
+            )
+        }
+    )
+
+
+def get_mappings_expanded(logger, extra_platform_data, variable_mappings, platform):
+    """Generate a dictionary of mappings between Platform group variables and external variables."""
```

Comment:
Could you please expand the typing and docstring for the arguments, so that it is easier to debug later? It is a little elusive at the moment which variable is of what type and what they present.

Reply:
Updated, elaborated the docstring for this method!